### PR TITLE
Don't crash on dead code in ICodeReader

### PR DESCRIPTION
**Pull request by lrytz**
Under `-Ybackend:GenBCode`, the scala compiler occasionally emits dead code, see discussion on PR #3726 for example.

Such a class file crashes ICodeReader. This was uncovered because under `-optimise`, the compiler switches to the ICode-GenASM pipeline, even if the `-Ybackend:GenBCode` flag is present.

There are multiple [tests that fail before](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1227/consoleFull) basically because the compiler emits inliner warnings ("could not inline") when the ICodeReader failed.

### CODE REVIEW

**Comment by retronym** [comment]
Maybe the test could mutate a valid class with ASM to inject dead frames.  That way the regular compiler is still in charge of signatures. 

Does icodereader also work for java classes? That would be simpler to craft in the test.

**Comment by lrytz** [comment]
I don't think we ever inline a java defined method.

Maybe it could work with an interpreter test, and we can change a class in memory, as you suggest.

File: test/files/run/repl-javap-app.scala [hunk 1]
```diff
@@ -7,4 +7,15 @@ object MyApp extends App {
...
+  override def welcoming = true
+
+  // The constant pool indices are not the same for GenASM / GenBCode, so
+  // replacing the exact numbers by XX.
+  lazy val hasConstantPoolRef = """(.*)(#\d\d)(.*)""".r
+
+  override def normalize(s: String) = s match {
+    case hasConstantPoolRef(start, ref, end) => start + "#XX" + end
+    case _ => super.normalize(s)
+  }
```
Comment (som-snytt) [inline test/files/run/repl-javap-app.scala:20]: Nice use of hook.

**Comment by lrytz** [comment]
Fixed failing tests, added a test for the fix in ICodeReader. review by @retronym 
`-Ybackend:GenBCode` build [passed](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1229/consoleFull)
cc @dragos

**Comment by retronym** [comment]
LGTM. I like the test.

**Pull request merged**